### ISSUE TITLE
Authorization Error when using AIS (Azure Identity Service) instead of a connection string.

### ISSUE THREAD

**Issue by Awarty** [Azure/iot-hub-device-update#459]
Hello!

I’m trying to get ADU to use Identity Service for connection to Azure. I have followed the guides below:

- [https://learn.microsoft.com/en-us/azure/iot-hub-device-update/device-update-agent-provisioning](https://learn.microsoft.com/en-us/azure/iot-hub-device-update/device-update-agent-provisioning)
- [https://learn.microsoft.com/en-us/azure/iot-hub-device-update/device-update-raspberry-pi](https://learn.microsoft.com/en-us/azure/iot-hub-device-update/device-update-raspberry-pi)

Setting up just ADU with a specified connection string in the du-config.json file works. When I tried to change to use AIS, instead it failed and got an error saying ‘Invalid connection info. [HealthCheck]’.

The full error from ‘AducIotAgent’:
<details>
<summary>Log 1</summary>

```
root@raspberrypi3:~# AducIotAgent 
2023-04-17T11:39:28.3253Z 4202[4202] [I] Agent (linux; 1.0.2) starting. [main]
2023-04-17T11:39:28.3254Z 4202[4202] [I] Git Info: main:2d1f0367 [main]
2023-04-17T11:39:28.3255Z 4202[4202] [I] Supported Update Manifest version: min: 4, max: 5 [main]
Error: Time:Mon Apr 17 11:39:28 2023 File:/usr/src/debug/azure-iot-sdk-c/1.0+gitAUTOINC+7329f70906-r0/git/c-utility/adapters/socketio_berkeley.c Func:initiate_socket_connection Line:353 Failure: connect failure 13.
Error: Time:Mon Apr 17 11:39:28 2023 File:/usr/src/debug/azure-iot-sdk-c/1.0+gitAUTOINC+7329f70906-r0/git/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:829 lookup_address_and_connect_socket failed
Error: Time:Mon Apr 17 11:39:28 2023 File:/usr/src/debug/azure-iot-sdk-c/1.0+gitAUTOINC+7329f70906-r0/git/deps/uhttp/src/uhttp.c Func:uhttp_client_open Line:1134 opening xio failed
2023-04-17T11:39:28.3262Z 4202[4202] [I] Failed to provision a connection string from eis, Failed with error EISErr_ConnErr on service EISService_IdentityService [GetConnectionInfoFromIdentityService]
2023-04-17T11:39:28.3263Z 4202[4202] [E] Invalid connection info. [HealthCheck]
2023-04-17T11:39:28.3265Z 4202[4202] [I] Health check failed. [HealthCheck]
2023-04-17T11:39:28.3266Z 4202[4202] [E] Agent health check failed. [main]
2023-04-17T11:39:28.3267Z 4202[4202] [I] Agent exited with code 1 [main]
2023-04-17T11:39:28.3267Z 4202[4202] [I] Agent is shutting down with signal 0. [ShutdownAgent]
2023-04-17T11:39:28.3268Z 4202[4202] [I] De-initializing command listener thread [UninitializeCommandListenerThread]
2023-04-17T11:39:28.3268Z 4202[4202] [I] ADUC agent stopping [AzureDeviceUpdateCoreInterface_Destroy]
```
</details>

I changed in the service file for ‘aziot-identityd’ to debug and got the following logs.
<details>
<summary>Log 2</summary>

```
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [INFO] - !!! authorization error
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::io] parsed 2 headers
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [INFO] - !!! authorization error
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [DBUG] - [hyper::proto::h1::io] parsed 2 headers
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - !!! authorization error
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [DBUG] - [hyper::proto::h1::io] parsed 2 headers
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [INFO] - !!! authorization error
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:15:10 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:10Z [DBUG] - [hyper::proto::h1::conn] read eof
```
</details>

I also found the following links containing information about configuring identity service.

- [https://azure.github.io/iot-identity-service/develop-an-agent.html](https://azure.github.io/iot-identity-service/develop-an-agent.html)
- [https://github.com/Azure/iot-hub-device-update/issues/414](https://github.com/Azure/iot-hub-device-update/issues/414)

Following the steps mentioned did not solve it for me but the different verification methods specified was helpful. I tried both `curl --unix-socket '/run/aziot/identityd.sock' 'http://identityd.sock/identities/identity?api-version=2020-09-01'` and `curl -XGET --unix-socket /run/aziot/identityd.sock http:/foo/identities/identity?api-version=2020-09-01`. Which resulted in the following response for both:

```
{"type":"aziot","spec":{"hubName":"XXXX","gatewayHost":"XXXX","deviceId":"XXXX","auth":{"type":"sas","keyHandle":"XXXX"}}}
```

And the logs from the identity service, where you can see a response with code 200. Which is the request sent by curl.
<details>
<summary>Log 3</summary>

```
Apr 17 12:14:49 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:49Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:14:49 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:49Z [INFO] - !!! authorization error
Apr 17 12:14:49 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:49Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:14:49 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:49Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:14:49 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:49Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::io] parsed 3 headers
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [INFO] - <-- GET /identities/identity?api-version=2020-09-01 {"host": "foo", "user-agent": "curl/7.69.1", "accept": "*/*"}
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::io] flushed 71 bytes
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::io] parsed 3 headers
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::conn] incoming body is content-length (240 bytes)
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::conn] incoming body completed
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::client::pool] pooling idle connection for ("http", keyd.sock)
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [INFO] - --> 200 {"content-type": "application/json"}
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::io] flushed 538 bytes
Apr 17 12:14:51 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:51Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [DBUG] - [hyper::proto::h1::io] parsed 2 headers
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [INFO] - !!! authorization error
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:14:54 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:54Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::io] parsed 2 headers
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [INFO] - !!! authorization error
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [INFO] - --> 401 {"content-type": "application/json"}
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::io] flushed 151 bytes
Apr 17 12:14:59 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:14:59Z [DBUG] - [hyper::proto::h1::conn] read eof
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [DBUG] - [hyper::proto::h1::io] parsed 2 headers
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [DBUG] - [hyper::proto::h1::conn] incoming body is empty
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - !!! authorization error
Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - --> 401 {"content-type": "application/json"}
```
</details>

In short, my problem is that when I tried to change from a connection string in du-config.json to using AIS. It fails to connect to Azure after I have configured AIS using `aziotctl config mp -c "<conn_string>"`. Is there anything I’m missing when configuring ADU and AIS? If you need any additional information please specify so I can add it.

**Comment by jw-msft** [comment]
> Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - <-- GET http://foo/identities/identity?api-version=2020-09-01 {"host": "/run/aziot/identityd.sock:0", "content-length": "0"}
> Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - !!! authorization error
> Apr 17 12:15:04 raspberrypi3 aziot-identityd[4783]: 2023-04-17T12:15:04Z [INFO] - --> 401 {"content-type": "application/json"}

The 401 authorization err from "Log 2" / "Log3" indicates an issue with user/group membership or .toml config file permission bits, so the process is not authorized to call into the identity service's Unix Domain Socket.

As per the steps outlined in client-authorization section of [iot-identity-service develop-an-agent.html](https://azure.github.io/iot-identity-service/develop-an-agent.html#client-authorization), can you confirm that the `adu` unix user is a member of the following unix groups?
- aziotid
- aziotcs
- aziotks

Also, ensure that the azure identity .toml config file has user and group owners as `aziotid` and `aziotid`, respectively.

See the below Debian postint script's `register_adu_user_and_process_with_eis()` function as a reference:
https://github.com/Azure/iot-hub-device-update/blob/main/packages/debian/postinst#L240

**Comment by Awarty** [comment]
Hi!

I have looked into your comment and found that I probably missed setting some user permissions on the image. So now I have it working correctly. Somewhat related to my issue was a problem with AducIotAgent version 1.0.0, which did not work correctly. Upgrading to 1.0.2 fixed these issues for me. 

Thanks for your fast and clear response.
